src/Equations/elastic/Kernels/GravitationalFreeSurfaceBC.h

```diff
@@ -127,103 +112,82 @@ class GravitationalFreeSurfaceBc {
-      auto[dEtaStorage, dEtaSize] = du.getSubvector(1);
-      assert(dEtaSize == init::faceDisplacement::size());
-      auto dEta = init::faceDisplacement::view::create(dEtaStorage);
-
-      // Unpack u
-      auto[etaStorage, etaSize] = u.getSubvector(1);
-      assert(etaSize == init::faceDisplacement::size());
-      auto eta = init::faceDisplacement::view::create(etaStorage);
-
-      constexpr int pIdx = 0;
-      constexpr int uIdx = 6;
-
-      dEta.setZero();
-      dEtaIntegrated.setZero();
+      projectKernel.execute(order - 1, faceIdx);
+
+      factorEvaluated *= deltaT / (1.0 * order);
+      factorInt *= deltaTInt / (order + 1.0);
+
+#pragma omp simd
       for (unsigned int i = 0; i < nodal::tensor::nodes2D::Shape[0]; ++i) {
-        const double rho = materialData.local.rho;
-        const double g = getGravitationalAcceleration(); // [m/s^2]
-        double pressureAtBnd = 0;
-        if (faceType == FaceType::freeSurfaceGravity) {
-          pressureAtBnd = -1 * rho * g * eta(i, 0);
-        }
-#ifdef USE_ELASTIC
-        const double Z = std::sqrt(materialData.local.lambda * rho) ;
-        bool isAcoustic = std::abs(materialData.local.mu) < 1e-15;
-#else
-        const real Z = 0.0; // Sets penalty term effectively to zero
-        bool isAcoustic = false;
-#endif
+        // Derivatives of interior variables
+        constexpr int pIdx = 0;
+        constexpr int uIdx = 6;
 
-        // dH/dt = etaN,
-        // dEtaN/dt =  u_r - 1/Z ( p_r - pressureAtBnd ),
         const auto uInside = dofsFaceNodal(i, uIdx + 0);
         const auto vInside = dofsFaceNodal(i, uIdx + 1);
         const auto wInside = dofsFaceNodal(i, uIdx + 2);
         const auto pressureInside = dofsFaceNodal(i, pIdx);
-        if (isAcoustic
-            && (faceType == FaceType::freeSurface || faceType == FaceType::freeSurfaceGravity)) {
-          dEta(i, 0) = uInside + 1 / Z * (pressureAtBnd - pressureInside);
-        } else {
-          // If not on acoustic boundary, just use values inside
-          dEta(i, 0) = uInside;
-        }
-        dEta(i, 1) = vInside;
-        dEta(i, 2) = wInside;
-        dEtaIntegrated(i) = eta(i, 0);
-      }
-    };
-
-    constexpr auto integratedEtaSize = init::averageNormalDisplacement::size();
-    constexpr auto etaSize = init::faceDisplacement::size();
-
-    auto curValue = ode::ODEVector{{integratedDisplacementNodal, rotatedFaceDisplacement},
-                                   {integratedEtaSize,           etaSize}};
 
-    // Apply initial condition to integrated displacement (start from 0 each PDE timestep)
-    std::fill_n(integratedDisplacementNodal, integratedEtaSize, 0.0);
+#ifdef USE_ELASTIC
+        const double curCoeff = uInside - (1.0/Z) * (rho * g * prevCoefficients[i] + pressureInside);
+        // Basically uInside - C_1 * (c_2 * prevCoeff[i] + pressureInside)
+        // 2 add, 2 mul = 4 flops
```

Comment:
Here you rely on the compiler to precompute `1/Z` and `rho*g`, otherwise I count 6 flops.

Reply:
Yes, but the compiler figures this out. At least it should.